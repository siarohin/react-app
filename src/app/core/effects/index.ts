import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
