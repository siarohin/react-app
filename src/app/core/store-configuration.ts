import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
