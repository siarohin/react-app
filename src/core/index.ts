export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
