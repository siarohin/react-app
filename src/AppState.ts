import values from "lodash/values";

import { FILTER_GENRES, FilterOptions, MoviesModels } from "./core";
import { IAppState } from "./IAppState";

/**
 * App state
 */
export const AppState: IAppState = {
  movieDialog: {
    dialogSettings: {
      title: "",
      values: {} as MoviesModels.IMovie
    },
    open: false
  },
  genres: {
    all: [...FILTER_GENRES],
    selected: ""
  },
  sortingOptions: {
    options: values(FilterOptions),
    selected: ""
  }
};
