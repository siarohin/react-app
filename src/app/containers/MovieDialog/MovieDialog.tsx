import React, { ComponentType, ReactElement } from "react";
import { useHistory } from "react-router-dom";
import { connect } from "react-redux";

import { DialogTitle, SharedModels, MoviesModels, MoviesActions, State, UserPreferencesModels } from "../../core";
import { IMovieDialog } from "../../shared";
import { getPath } from "../../utils";
import { DeleteMovieDialog, UpsertMovieDialog } from "../../components";

const MovieDialog = (
  props: Omit<IMovieDialog, "onSubmit"> & SharedModels.IDispatchAction & { searchValue: string }
): ReactElement<typeof DeleteMovieDialog> | ReactElement<typeof UpsertMovieDialog> => {
  const { open, onClose, dialogSettings, genres, searchValue, dispatch } = props;
  const history = useHistory();
  const isActionDelete: boolean = dialogSettings?.title === DialogTitle.DELETE;

  const handleSubmit = (movie: MoviesModels.IMovie): void => {
    const isActionCreate: boolean = isNaN(Number(movie.id));

    if (isActionCreate) {
      dispatch(MoviesActions.createMovie({ movie }));
    } else {
      dispatch(MoviesActions.updateMovie({ movie }));
    }

    onClose();
  };

  const handleDelete = (movie: MoviesModels.IMovie): void => {
    const value: string = searchValue?.trim();

    dispatch(MoviesActions.deleteMovie({ movie }));
    history.push(getPath(value));

    onClose();
  };

  if (isActionDelete) {
    return <DeleteMovieDialog open={open} onClose={onClose} onDelete={handleDelete} dialogSettings={dialogSettings} />;
  } else {
    return (
      <UpsertMovieDialog
        genres={genres}
        open={open}
        onClose={onClose}
        onSubmit={handleSubmit}
        dialogSettings={dialogSettings}
      />
    );
  }
};

const mapStateToProps = (state: State): { genres: UserPreferencesModels.IGenres; searchValue: string } => {
  return {
    genres: state.userPreferences.genres,
    searchValue: state.userPreferences.search.selected
  };
};

export default connect(mapStateToProps)(MovieDialog as ComponentType<any>);
