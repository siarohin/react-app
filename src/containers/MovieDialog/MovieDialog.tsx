import React, { ComponentType, ReactElement } from "react";
import { connect } from "react-redux";

import { DialogTitle, SharedModels, MoviesModels, MoviesActions, State, UserPreferencesModels } from "../../core";
import { DeleteMovieDialog, UpsertMovieDialog } from "../../components";
import { IMovieDialog } from "../../models";

const MovieDialog = (
  props: Omit<IMovieDialog, "onSubmit"> & SharedModels.IDispatchAction
): ReactElement<typeof DeleteMovieDialog> | ReactElement<typeof UpsertMovieDialog> => {
  const { open, onClose, dialogSettings, genres, dispatch } = props;
  const isActionDelete: boolean = dialogSettings?.title === DialogTitle.DELETE;

  const handleSubmit = (movie: MoviesModels.IMovie): void => {
    const isActionCreate: boolean = isNaN(Number(movie.id));

    if (isActionCreate) {
      dispatch(MoviesActions.createMovie({ movie }));
    } else {
      dispatch(MoviesActions.updateMovie({ movie }));
    }

    onClose();
  };

  const handleDelete = (movie: MoviesModels.IMovie): void => {
    dispatch(MoviesActions.deleteMovie({ movie }));
    onClose();
  };

  if (isActionDelete) {
    return <DeleteMovieDialog open={open} onClose={onClose} onDelete={handleDelete} dialogSettings={dialogSettings} />;
  } else {
    return (
      <UpsertMovieDialog
        genres={genres}
        open={open}
        onClose={onClose}
        onSubmit={handleSubmit}
        dialogSettings={dialogSettings}
      />
    );
  }
};

const mapStateToProps = (state: State): Partial<UserPreferencesModels.IUserPreferencesState> => {
  return { genres: state.userPreferences.genres };
};

export default connect(mapStateToProps)(MovieDialog as ComponentType<any>);
