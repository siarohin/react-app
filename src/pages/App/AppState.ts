import { MoviesModels } from "../../core";
import { IAppState } from "./IAppState";

/**
 * App state
 */
export const AppState: IAppState = {
  movieDialog: {
    dialogSettings: {
      title: "",
      values: {} as MoviesModels.IMovie
    },
    open: false
  }
};
