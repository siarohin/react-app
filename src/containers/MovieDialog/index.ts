import MovieDialog from "./MovieDialog";

export { MovieDialog };
