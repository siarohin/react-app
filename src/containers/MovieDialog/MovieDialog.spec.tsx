import React from "react";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";
import noop from "lodash/noop";

import { State } from "../../core";
import { AppState } from "../../AppState";
import MovieDialog from "./MovieDialog";

const mockStore = configureStore([]);

describe("Containers.MovieDialog.MovieDialog: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  const { genres, movieDialog } = AppState;

  beforeEach(() => {
    store = mockStore({
      movieList: {
        movies: [],
        isLoading: false
      }
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <MovieDialog
          genres={genres.all}
          dialogSettings={movieDialog.dialogSettings}
          open={movieDialog.open}
          onClose={() => noop}
        />
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
