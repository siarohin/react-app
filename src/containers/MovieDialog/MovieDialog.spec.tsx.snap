// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.MovieDialog.MovieDialog:  should render component 1`] = `null`;
